### PR TITLE
0.1.8

### PR DESCRIPTION
**Pull request by btkostner**
Road to 0.2.0!

This is rather big but it works :D

Includes:
- Koa, a new framework by the creators of express using new technology like es7 await/async
- Less model code, cause simplicity
- Hopefully dryer code, cause kiss
- Modularity, cause all the reasons above
- Difference between code and appHooks, cause easier PRs (hopefully)
- Documentation, cause people love reading about these types of things

Things to do before 0.2.0:
- ~~Update appHook tests to new format~~ (#122)
- ~~View engine for appHook issues~~ (#121)
- Pushing packages to repo (#125)
- Way to get new releases (GitHub hooks? or just page to fetch releases? probably the second for now)

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# EditorConfig is awesome: http://EditorConfig.org
```
Comment (aroman) [inline .editorconfig]: cool, hadn't seen this before! Just installed the Atom plugin :)

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
+# Houston
+**Prepare for liftoff.**
+
+Firstly, thank you for taking the time to contribute to Houston :+1:
+
+Houston would be but a glimmering dream if it were not for the astounding
+contributions make by third-party developers. To keep contribution as easy as
```
Comment (aroman) [inline CONTRIBUTING.md]: made

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
+# Houston
+**Prepare for liftoff.**
+
+Firstly, thank you for taking the time to contribute to Houston :+1:
+
+Houston would be but a glimmering dream if it were not for the astounding
```
Comment (aroman) [inline CONTRIBUTING.md]: dope

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
+# Houston
+**Prepare for liftoff.**
+
+Firstly, thank you for taking the time to contribute to Houston :+1:
```
Comment (aroman) [inline CONTRIBUTING.md]: dunno if "firstly" is really needed here, especially since there's no "second" :stuck_out_tongue:

File: LICENSE.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# The MIT License (MIT)
+
+### Copyright © 2016 elementary & [The Web Team](https://github.com/elementary/houston/graphs/contributors)
```
Comment (aroman) [inline LICENSE.md]: maybe "& The Houston Contributors?"

cc @danrabbit

File: app.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+export const Helpers = require('./helpers')
+
+// Application configuration loading from config.js
+try {
+  Fs.statSync('./config.js')
+} catch (err) {
+  console.log('It seems like you have not taken the time to setup Houston yet.')
+  console.log('Please use the example configuration file we have provided you')
+  console.log('to setup Houston.')
+  throw new Error('failed to setup Houston')
```
Comment (aroman) [inline app.js]: i think a better error would be 'config.js file missing', which is more accurate.

File: app.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+})
+
+app.db.connection.once('open', function () {
+  app.log.info('Connected to database')
+})
+
+export const Db = app.db
+
+// Export an amazing request library
+export const Request = Super
```
Comment (aroman) [inline app.js]: +1, i love superagent

File: appHooks/appHook.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  error (msg) {
+    this.errors.push(msg)
+  }
+
+  warn (msg) {
+    this.warnings.push(msg)
+  }
+
+  update (obj) {
+    this.information = _.extend(this.information, obj)
```
Comment (aroman) [inline appHooks/appHook.js]: in ES6, you can use `Object.assign()`. would also mean you can ditch the _ require in this file :)

File: appHooks/appHook.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+      errors: this.errors.length,
+      warnings: this.warnings.length,
+      information: this.information,
+      issue: (this.errors.length > 0 || this.warnings.length > 0) ? this.issue() : null
+    }
+  }
+
+  issue () {
+    // TODO: add a templating language
+    let template = fs.readFileSync(`${this.path}/${this.mark}`, 'utf8')
```
Comment (aroman) [inline appHooks/appHook.js]: why `readFileSync` — especially since you've already promisify-ed `fs`?

File: appHooks/apphub/prebuild.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      hook.error('parse');
+      return Promise.reject();
+    }
+
+    return Promise.resolve();
+  });
+});
+
+hook.question(release => {
+  let apphub = this.metadata;
```
Comment (aroman) [inline appHooks/apphub/prebuild.js]: `const`

File: appHooks/apphub/prebuild.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    hook.error('price');
+  } else if (apphub.priceUSD != null) {
+    hook.update({priceUSD: apphub.priceUSD});
+  }
+
+  return Promise.resolve();
+});
+
+hook.question(release => {
+  let apphub = this.metadata;
```
Comment (aroman) [inline appHooks/apphub/prebuild.js]: `const`

File: appHooks/desktop/prebuild.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Prebuild parse and test application desktop file
+ */
+import Promise from 'bluebird';
+import Hubkit from 'hubkit';
+import AppHook from '~/appHooks/appHook';
+
+const gh = new Hubkit();
+let hook = new AppHook('desktop', __dirname);
```
Comment (aroman) [inline appHooks/desktop/prebuild.js]: const?

File: appHooks/apphub/prebuild.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+hook.question(release => {
+  const application = release.ownerDocument();
+
+  return gh.request(`GET /repos/${application.github.fullName}/contents/.apphub?ref=${release.github.tag}`, {
+    token: application.github.APItoken,
+  })
+  .then(apphubData => {
+    try {
+      const apphubBase = new Buffer(apphubData.content, 'base64');
+      hook.meta(JSON.parse(apphubBase.toString()));
```
Comment (aroman) [inline appHooks/apphub/prebuild.js]: where is the `meta` method defined? what does it do?

File: appHooks/icon/prebuild.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+  return gh.request(`GET /repos/${application.github.fullName}/contents/icons/64/${application.github.name}.svg?ref=${release.github.tag}`, {
+    token: application.github.APItoken,
+  })
+  .then(icon => {
+    hook.update({icon});
+  })
+  .catch(err => {
+    if (err.status === 404) {
+      hook.error(64)
```
Comment (aroman) [inline appHooks/icon/prebuild.js]: what's error 64?

File: appHooks/index.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+import Promise from 'bluebird'
+import _ from 'lodash'
+
+import { Log } from '~/app'
+
+let fs = Promise.promisifyAll(require('fs'))
+
+function getHooks (lvl) {
+  const level = lvl.toLowerCase()
+  return Promise.all(fs.readdirSync(__dirname))
```
Comment (aroman) [inline appHooks/index.js]: same comment as before — why sync if you've promisified?

File: appHooks/index.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+import Promise from 'bluebird'
+import _ from 'lodash'
+
+import { Log } from '~/app'
+
+let fs = Promise.promisifyAll(require('fs'))
+
+function getHooks (lvl) {
```
Comment (aroman) [inline appHooks/index.js]: this is a complex function, would be nice to have a little docstring header, at least explaining what the arguments are and what it returns.

File: appHooks/index.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    .catch(() => false)
+  })
+  .map(dir => {
+    return require(`${__dirname}/${dir}/${level}.js`)
+  })
+}
+
+export async function run (data) {
+  let hooks = await getHooks(data.status)
+  let tests = []
```
Comment (aroman) [inline appHooks/index.js]: why `const tests = hooks.map(...)`?

File: appHooks/index.js [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+  return Promise.all(tests)
+  .catch(err => Log.error(err))
+  .then(pkg => {
+    let obj = {errors: 0, warnings: 0, information: {}, issues: []}
+
+    for (let i in pkg) {
+      obj.errors = pkg[i].errors + obj.errors
+      obj.warnings = pkg[i].warnings + obj.warnings
+      obj.information = _.extend(obj.information, pkg[i].information)
```
Comment (aroman) [inline appHooks/index.js]: `Object.assign()`

File: appHooks/index.js [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+      if (pkg[i].issue != null) obj.issues.push(pkg[i].issue)
+    }
+
+    return obj
+  })
+  .then(pkg => {
+    return _.extend({
```
Comment (aroman) [inline appHooks/index.js]: `Object.assign()`. then drop the lodash require

File: core/controller/hook/github.js [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @exports {Object} default - Koa router
+ */
+
+import Router from 'koa-router'
+
+import { Config, Log } from '~/app'
+
+// TODO: GitHub hook integration
+let route = new Router({
```
Comment (aroman) [inline core/controller/hook/github.js]: this is a pretty sweet API. hadn't really messed with Koa much, but very cool!

File: core/controller/hook/jenkins.js [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+  prefix: '/hook/jenkins/:key'
+})
+
+route.param('key', async (key, ctx, next) => {
+  if (!Config.jenkins) {
+    Log.debug('Jenkins is disabled but someone tried to access Jenkins hook')
+    return ctx.throw('Jenkins is on vacation right now', 404)
+  } else if (key !== Config.jenkins.public) {
+    Log.debug('Someone tried to connect to Jenkins with an incorrect key')
+    return ctx.throw('Nobody can replace Mr. Jenkins', 404)
```
Comment (aroman) [inline core/controller/hook/jenkins.js]: lol

File: core/controller/hook/jenkins.js [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  await next()
+})
+
+route.post('/', async ctx => {
+  if (ctx.request.body === {}) {
+    Log.debug('Incorrect data was sent to Jenkins hook')
```
Comment (aroman) [inline core/controller/hook/jenkins.js]: maybe "body was empty", more specifically...?

File: core/controller/hook/jenkins.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+  const jenkins = ctx.request.body.build
+  const build = await Build.findById(jenkins.parameters.BUILD)
+
+  if (build == null) {
+    return ctx.throw('No Build found', 404)
+  }
+
+  let status = 'QUEUE'
+  if (jenkins.phase === 'QUEUED') status = 'QUEUE'
```
Comment (aroman) [inline core/controller/hook/jenkins.js]: have you considered using ES6 symbols for this, assuming the pollyfills/babel support it? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol

File: core/controller/project.js [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+})
+
+route.get('/init', async (ctx, next) => {
+  const status = await ctx.project.getStatus()
+  .catch(err => ctx.throw({
+    message: 'Unable to get project status',
+    error: err
+  }, 500))
+
+  if (status !== 'NEW') return ctx.throw('The project is already in standby', 400)
```
Comment (aroman) [inline core/controller/project.js]: again, i would be more specific with the error message here. technically, you just know the status isn't `'NEW'`

File: core/controller/project.js [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+    return ctx.throw('Release is not awaiting review', 400)
+  }
+
+  let newStatus = 'REVIEW'
+  if (ctx.params.fate === 'yes') {
+    newStatus = 'FINISH'
+  } else if (ctx.params.fate === 'no') {
+    newStatus = 'FAIL'
+  } else {
+    return ctx.throw(`${ctx.project.name}'s fate is binary'`, 400)
```
Comment (aroman) [inline core/controller/project.js]: what's this a reference to? #outOfTheLoop

File: core/controller/root.js [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+
+route.get('/', async (ctx) => {
+  return ctx.render('index', { hideUser: true })
+})
+
+route.get('/dashboard', IsRole('USER'), async (ctx, next) => {
+  const projects = await GetProjects(ctx.user.github.access)
+  .map(async project => {
+    Log.silly(`Upserting ${project.github.owner}/${project.github.name} project`)
+    let dbProject = await Project.findOne({ 'github.id': project.github.id })
```
Comment (aroman) [inline core/controller/root.js]: `const`

File: core/controller/root.js [hunk 27]
```diff
@@ -0,0 +1,69 @@
+/**
+ * core/controller/root.js
+ * Handles the dashboard and static pages
+ *
+ * @exports {Object} default - Koa router
+ */
+
+import Router from 'koa-router'
```
Comment (aroman) [inline core/controller/root.js]: why are there no semicolons in this file, btw?

File: core/controller/root.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+route.get('/dashboard', IsRole('USER'), async (ctx, next) => {
+  const projects = await GetProjects(ctx.user.github.access)
+  .map(async project => {
+    Log.silly(`Upserting ${project.github.owner}/${project.github.name} project`)
+    let dbProject = await Project.findOne({ 'github.id': project.github.id })
+    if (dbProject != null) return dbProject
+
+    Log.silly(`Creating a new project for ${project.github.owner}/${project.github.name}`)
+    // TODO: Verify only owners? of projects can create it
+    return Project.create(_.extend({
```
Comment (aroman) [inline core/controller/root.js]: `Object.assign()`

File: core/controller/hook/github.js (hunk 20 above)
Comment (btkostner) [inline core/controller/hook/github.js]: After using it a while, I really enjoy how simple it is to lay down requirements for different paths. For instance, being able to make the GitHub hook only respond to this path, automatically verify the key, and then pass along a database object to the next route. Best part is that I have yet to get into some weird circumstance where it doesn't act how I want it to.

File: core/model/cycle.js [hunk 29]
```diff
@@ -0,0 +1,205 @@
...
+
+// TODO: make a pkg to sent to slave script
+CycleSchema.methods.spawn = async function () {
+  return Promise.all([
+    this.getProject(),
+    this.getRelease(),
+    this.getRepo(),
+    this.getTag()
+  ])
+  .then(([project, release, repo, tag]) => {
```
Comment (aroman) [inline core/model/cycle.js]: this whole little section is sooo ES6-y :+1:

File: core/model/cycle.js [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+  })
+}
+
+CycleSchema.methods.build = async function () {
+  const cycle = this
+  const project = await cycle.getProject()
+
+  if (project == null) return project.reject('Unable to find project')
+
+  let builds = []
```
Comment (aroman) [inline core/model/cycle.js]: why not `map`?

File: core/model/index.js [hunk 31]
```diff
@@ -0,0 +1,10 @@
+/**
+ * core/model/index.js
+ * Exports all models for convenience
+ */
+
+export * as Build from './build'
```
Comment (aroman) [inline core/model/index.js]: cooooool

File: core/controller/project.js (hunk 25 above)
Comment (btkostner) [inline core/controller/project.js]: I can't remember. Not even sure if it is a reference (Google didn't have anything). It just felt like something funny that actually kinda made some sense.

File: core/controller/root.js (hunk 27 above)
Comment (btkostner) [inline core/controller/root.js]: using standard to lint everything

File: entry.js [hunk 32]
```diff
@@ -0,0 +1,27 @@
+/**
+ * entry.js
+ * Simple wrapper for babel
+ */
+
+require('babel-register')
+require('babel-polyfill')
+
+var currentIndex = process.argv.indexOf(__filename)
```
Comment (aroman) [inline entry.js]: what's all this about?

File: master.js [hunk 33]
```diff
@@ -0,0 +1,112 @@
+/**
+ * master.js
+ * Starts Houston's Web Interface
+ *
+ * @exports {Object} App
+ */
+
+import Co from 'co'
+import Convert from 'koa-convert'
```
Comment (aroman) [inline master.js]: sweet

File: master.js [hunk 34]
```diff
@@ -0,0 +1,112 @@
...
+Passport.Setup(App)
+
+App.use(Passport.Route.routes(), Passport.Route.allowedMethods())
+
+// Load Houston core files
+let routes = Helpers.FlattenObject(Controller, { Route: 'object' })
+
+for (let key in routes) {
+  let router = routes[key].Route
+  let path = router.opts.prefix || '/'
```
Comment (aroman) [inline master.js]: consts

File: master.js [hunk 35]
```diff
@@ -0,0 +1,112 @@
...
+Log.info(`Loaded ${Helpers.ArrayString('Controller', routes)}`)
+
+// 404 page
+App.use(async (ctx, next) => {
+  await ctx.render('error', { message: 'It seems you stuck the landing. World not found.' })
+})
+
+// Error logging
+App.on('error', function (error, ctx, next) {
+  if (App.env === 'test') return
```
Comment (aroman) [inline master.js]: isn't this the opposite of what you want? this reads to me like in 'test' mode, there is no verbose logging, but in non-test mode, there is

File: master.js [hunk 36]
```diff
@@ -0,0 +1,112 @@
...
+// Error pages
+App.use(async (ctx, next) => {
+  try {
+    await next()
+  } catch (error) {
+    ctx.app.emit('error', error, ctx)
+    if (error.expose) {
+      await ctx.render('error', { message: error.message })
+    } else {
+      await ctx.render('error', { message: 'Houston, we have a problem' })
```
Comment (aroman) [inline master.js]: http://keeba.jbha.org/unsupported

File: master.js [hunk 37]
```diff
@@ -0,0 +1,112 @@
...
+App.use(Parser())
+App.keys = [ Config.server.secret ]
+App.use(Convert(Session(App)))
+
+Passport.Setup(App)
+
+App.use(Passport.Route.routes(), Passport.Route.allowedMethods())
+
+// Load Houston core files
+let routes = Helpers.FlattenObject(Controller, { Route: 'object' })
```
Comment (aroman) [inline master.js]: what is this section, exactly?

File: package.json [hunk 38]
```diff
@@ -1,38 +1,64 @@
...
+  },
+  "standard": {
+    "parser": "babel-eslint"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/elementary/houston.git"
+  },
+  "author": "elementary",
+  "license": "MIT",
```
Comment (aroman) [inline package.json]: are we sure we want to MIT license?

File: README.md [hunk 39]
```diff
@@ -1,113 +1,148 @@
...
-* Needs Release: This repo needs a new git tag that is not already associated with a build and with a suitable changelog before it can be published
-* Publish: The repo has a git tag that doesn't already have an associated build
-* In Progress: The repo is currently being tested or built.
-* Failure: The repo was not succesfully published as a debian package due to build failure or test suite failure. At this stage, houston will file issues with more detail into the repo's GitHub issues tracker
-* Needs Review: The git repo was succesfully built into a debian package and requires a human to move it from the "testing" deb repo to the "stable" deb repo. After this task is completed, start over with "Needs Release" status.
 
 ## Architecture Overview
+
 * HTML/CSS/JS
 * [Handlebars](http://handlebarsjs.com/) templates
```
Comment (aroman) [inline README.md]: now Jade instead, right?

File: core/controller/hook/jenkins.js [hunk 40]
```diff
@@ -0,0 +1,59 @@
+/**
```
Comment (aroman) [inline core/controller/hook/jenkins.js]: so really all of Jenkins is being to be replaced by the socketio-powered slave system, right?

File: core/controller/root.js (hunk 27 above)
Comment (aroman) [inline core/controller/root.js]: then get rid of them everywhere else, i think. but don't do that now, do it in a separate PR that is only changing the semicolons.

File: LICENSE.md (hunk 5 above)
Comment (danrabbit) [inline LICENSE.md]: "elementary" should probably be "elementary LLC". I'm unsure about the other part. I don't know if those things hold up in court of if there's a way to do non-legal entities like that

File: entry.js (hunk 32 above)
Comment (btkostner) [inline entry.js]: It's a generic script for babel polyfill that I made. So you would do something like `node entry.js master.js` and tada, all of babel. Currently use it for the slave and master script. Keeps everything nice and fresh looking so we don't have server scripts being in regular ol boring javascript.

File: package.json (hunk 38 above)
Comment (aroman) [inline package.json]: @danrabbit @fabianthoma

File: master.js (hunk 35 above)
Comment (btkostner) [inline master.js]: the 'test' environment *will* be used for things like mocha tests when I get around to it. So when you run the testing script you won't see all the server information etc, just if the test passed or failed. In some cases, like if you are testing web endpoints, you don't necessarily want a 404 error appearing, because that's what you actually want to happen.

File: master.js (hunk 36 above)
Comment (btkostner) [inline master.js]: Totally wasn't copying that >.>

File: master.js (hunk 37 above)
Comment (btkostner) [inline master.js]: Ah yes. This flattens the whole `core/controller` tree so it's an array of route functions. This is because most routers (this one included) don't support using a deep nested object of routes. Basically this is what makes it possible to have nice organization in the controller file.

File: package.json (hunk 38 above)
Comment (btkostner) [inline package.json]: Good point. I copied the mvp licensing stuff for this so we should definitely check with the team on this.

File: master.js (hunk 35 above)
Comment (aroman) [inline master.js]: ah, I see. i assumed 'test' was like 'development', not referring to automated test suites

File: master.js (hunk 36 above)
Comment (aroman) [inline master.js]: i know, i wrote that a very long time ago :)

File: core/controller/hook/jenkins.js (hunk 40 above)
Comment (btkostner) [inline core/controller/hook/jenkins.js]: eventually yes, but for now its an mvp

File: .editorconfig (hunk 1 above)
Comment (btkostner) [inline .editorconfig]: It's SUPER convenient for multiple projects. Keeps you sane not having to change your editor each time you move projects

File: appHooks/icon/prebuild.js (hunk 14 above)
Comment (btkostner) [inline appHooks/icon/prebuild.js]: icon size

File: appHooks/apphub/prebuild.js (hunk 13 above)
Comment (btkostner) [inline appHooks/apphub/prebuild.js]: it's just a function to store information for the issue. Mostly stuff like function dumps and stuff

File: core/controller/root.js (hunk 27 above)
Comment (btkostner) [inline core/controller/root.js]: all the core files are linted. The only reason the appHook files were not is because I have yet to update them. There should be an open PR for that now

File: core/controller/hook/jenkins.js (hunk 23 above)
Comment (btkostner) [inline core/controller/hook/jenkins.js]: I actually haven't looked to much into the ES6 symbols. Reading a bit about them I fail to see how we would use them here. Care to elaborate?

File: core/controller/project.js (hunk 24 above)
Comment (btkostner) [inline core/controller/project.js]: Well if the project's status isn't 'NEW' then we can assume we already setup all of the hooks (eventually), GitHub label, and stuff we need to. As such, all I can think of to say is 'The project is already initialized'

File: package.json (hunk 38 above)
Comment (btkostner) [inline package.json]: I'll just remove the LICENSE part until we decide I guess?

**Comment by fabianthoma** [comment]
I guess once #125 & #123 are merged into this it is up for another review before merging into master?

**Comment by btkostner** [comment]
Probably a good idea to glance at it before merging, but it shouldn't have any major problems

**Comment by fabianthoma** [comment]
Since this branch is now so big and almost unreviewable, I'm going to merge it because it seems like all of the concerns @aroman had been addressed.
